### PR TITLE
fix: avoid copy suggestion when only config is skipped

### PR DESCRIPTION
**Pull request by sajayantony**
**What this PR does / why we need it**:
The recommendation to use ORAS copy will be skipped if the only layer not downloaded is the config. 

## When we try to download an image

image

## When there is an artifact with one layer and only empty config is skipped - 

image

### CODE REVIEW

File: cmd/oras/root/pull.go [hunk 1]
```diff
@@ -219,14 +228,16 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
 		var ret []ocispec.Descriptor
 		for _, s := range nodes {
 			if s.Annotations[ocispec.AnnotationTitle] == "" {
-				skippedLayers++
+				if s.Digest != ocispec.DescriptorEmptyJSON.Digest {
+					skippedLayers++
+				}
```
Comment (qweeah) [inline cmd/oras/root/pull.go:233]: Suggest revert the change since it may also silently skipped an unnamed layer with content `{}`. We can skip config blob in the above `if config != nil` block.
```suggestion
			skippedLayers++
```

File: cmd/oras/root/pull.go [hunk 2]
```diff
@@ -219,14 +228,16 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
 		var ret []ocispec.Descriptor
```
Comment (qweeah) [inline cmd/oras/root/pull.go:228]: We can change the above `nodes = append(nodes, *config)` to below
```golang
if config.Size != ocispec.DescriptorEmptyJSON.Size && config.Digest != ocispec.DescriptorEmptyJSON.Digest {
    nodes = append(nodes, *config)
}

```

Comment (qweeah) [reply cmd/oras/root/pull.go:230]: Added in https://github.com/oras-project/oras/pull/1164/commits/3ab0152c38be2bafac2cb985d8731a8a00d3ae54

File: cmd/oras/root/cp.go [hunk 3]
```diff
@@ -154,38 +154,45 @@ func doCopy(ctx context.Context, src oras.ReadOnlyGraphTarget, dst oras.GraphTar
 		return graph.Referrers(ctx, src, desc, "")
 	}
 
+	const (
+		Exists  = "Exists "
+		Copying = "Copying"
+		Copied  = "Copied "
+		Skipped = "Skipped"
```
Comment (Wwwsylvia) [inline cmd/oras/root/cp.go:161]: Do these variables need to be public? 
nit: Since these variables are passed to `PrintStatus()` as `status`, should we name them`statusExists`, `statusCopying`, etc.  for better readability?

File: cmd/oras/root/pull.go [hunk 4]
```diff
@@ -159,6 +159,15 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
 		}
 	}
 
+	const (
+		Downloading = "Downloading"
+		Pulled      = "Pulled     "
+		Processing  = "Processing "
+		Skipped     = "Skipped    "
+		Restored    = "Restored   "
+		Downloaded  = "Downloaded "
```
Comment (Wwwsylvia) [inline cmd/oras/root/pull.go:168]: Do we need to export these?

File: cmd/oras/root/push.go [hunk 5]
```diff
@@ -242,33 +242,41 @@ func doPush(dst oras.Target, pack packFunc, copy copyFunc) (ocispec.Descriptor,
 func updateDisplayOption(opts *oras.CopyGraphOptions, fetcher content.Fetcher, verbose bool, tracked track.GraphTarget) {
 	committed := &sync.Map{}
 
+	const (
+		Skipped   = "Skipped  "
+		Uploaded  = "Uploaded "
+		Exists    = "Exists   "
+		Uploading = "Uploading"
+		Uploaded2 = "Uploaded "
```
Comment (Wwwsylvia) [inline cmd/oras/root/push.go:250]: What's the difference between `Uploaded` and `Uploaded2`? 🧐

File: cmd/oras/root/pull.go (hunk 2 above)
Comment (sajayantony) [reply cmd/oras/root/pull.go:230]: Do you want to send the PR over this @qweeah ?

File: cmd/oras/root/cp.go [hunk 6]
```diff
@@ -154,38 +154,45 @@ func doCopy(ctx context.Context, src oras.ReadOnlyGraphTarget, dst oras.GraphTar
...
+	)
```
Comment (shizhMSFT) [inline cmd/oras/root/cp.go:162]: nit: it will be more readable if there is a `label` prefix.
```suggestion
	const (
		labelExists  = "Exists "
		labelCopying = "Copying"
		labelCopied  = "Copied "
		labelSkipped = "Skipped"
	)
```

Comment (shizhMSFT) [reply cmd/oras/root/cp.go:162]: Same suggestions for other files

File: cmd/oras/root/push.go (hunk 5 above)
Comment (shizhMSFT) [reply cmd/oras/root/push.go:250]: `Uploaded2` is never used and thus can be removed.

Comment (shizhMSFT) [reply cmd/oras/root/push.go:250]: I'm curious if this `const` block is somehow auto-generated or not?

File: cmd/oras/root/pull.go [hunk 7]
```diff
@@ -213,20 +222,24 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
 					config.Annotations[ocispec.AnnotationTitle] = configPath
 				}
 			})
-			nodes = append(nodes, *config)
+			if config.Size != ocispec.DescriptorEmptyJSON.Size || config.Digest != ocispec.DescriptorEmptyJSON.Digest || config.Annotations[ocispec.AnnotationTitle] != "" {
```
Comment (shizhMSFT) [inline cmd/oras/root/pull.go:225]: [`content.Equal()`](https://pkg.go.dev/oras.land/oras-go/v2/content#Equal) is a better method to compare two descriptors.

File: cmd/oras/root/pull.go [hunk 8]
```diff
@@ -213,20 +222,24 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
...
+				nodes = append(nodes, *config)
+			}
 		}
 
 		var ret []ocispec.Descriptor
 		for _, s := range nodes {
 			if s.Annotations[ocispec.AnnotationTitle] == "" {
-				skippedLayers++
+				if s.Digest != ocispec.DescriptorEmptyJSON.Digest || s.MediaType != ocispec.DescriptorEmptyJSON.MediaType || s.Size != ocispec.DescriptorEmptyJSON.Size {
```
Comment (shizhMSFT) [inline cmd/oras/root/pull.go:233]: [`content.Equal()`](https://pkg.go.dev/oras.land/oras-go/v2/content#Equal) is a better method to compare two descriptors.

Comment (shizhMSFT) [inline cmd/oras/root/pull.go:233]: Do we need to check here as the config has already been filtered out?

File: cmd/oras/root/pull.go [hunk 9]
```diff
@@ -213,20 +222,24 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
...
+					skippedLayers++
```
Comment (shizhMSFT) [inline cmd/oras/root/pull.go:234]: I just noticed that `skippedLayers` is not go-routine safe.
@qweeah Let's fix it in another PR.

File: cmd/oras/root/cp.go (hunk 3 above)
Comment (qweeah) [reply cmd/oras/root/cp.go:161]: According to the naming of progress output, those should be `statusXXX`

File: cmd/oras/root/pull.go (hunk 4 above)
Comment (qweeah) [reply cmd/oras/root/pull.go:168]: No, since different command contains different prompt and all prompts used by one command should be padding to the maximum length. E.g.  `Skipped` for `oras pull` should be `Skipped    `, but for `oras cp` should be `Skipped`

File: cmd/oras/root/pull.go (hunk 7 above)
Comment (qweeah) [reply cmd/oras/root/pull.go:225]: No, it cannot be used here since the comparison should skip `mediaType`

File: cmd/oras/root/cp.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/root/cp.go:162]: How about `promptXXX`? It's used as status prompt in progress output

File: cmd/oras/root/pull.go (hunk 9 above)
Comment (qweeah) [reply cmd/oras/root/pull.go:234]: We can change it to [atomic.Bool](https://pkg.go.dev/sync/atomic#Bool) since the returned int value is not actually used.

File: cmd/oras/root/pull.go (hunk 8 above)
Comment (qweeah) [reply cmd/oras/root/pull.go]: No we don't, if the config is already an empty descriptor, it's not in the nodes.

**Review by shizhMSFT — APPROVED** [review]
LGTM
